File: frontend/src/api/vehicles.ts
import { apiClient } from './client'
import type { Vehicle, VehicleDetails } from '../types/vehicle'

type ApiVehicle = {
  id: string
  slug: string
  title: string
  make: string
  model: string
  trim: string | null
  year: number
  status: string
  stock_number: string | null
  vin: string | null
  price: number
  mileage: number
  body_type: string
  transmission: string | null
  drivetrain: string | null
  engine: string | null
  exterior_color: string | null
  interior_color: string | null
  short_description: string
  description?: string
  images: string[]
  images_total: number
  features: string[]
  specs: Record<string, string>
  details?: {
    info?: { label: string; value: string }[]
    sections?: { title: string; content: string }[]
    raw_description?: string
  }
  featured: boolean
}

export type VehicleListParams = {
  page?: number
  pageSize?: number
  make?: string
  model?: string
  yearFrom?: number
  yearTo?: number
  bodyType?: string
  transmission?: string
  drivetrain?: string
  color?: string
  featured?: boolean
  priceMin?: number
  priceMax?: number
  mileageMin?: number
  mileageMax?: number
  sort?: 'year_desc' | 'price_asc' | 'price_desc' | 'mileage_asc'
}

export type VehicleFilters = {
  makes: string[]
  models: string[]
  years: number[]
  bodyTypes: string[]
  transmissions: string[]
  drivetrains: string[]
  colors: string[]
  statuses: string[]
  prices: number[]
  priceMin: number | null
  priceMax: number | null
  mileageMin: number | null
  mileageMax: number | null
}

type ApiVehicleListResponse = {
  items: ApiVehicle[]
  total: number
  page: number
  page_size: number
}

type ApiVehicleFilters = {
  makes: string[]
  models: string[]
  years: number[]
  body_types: string[]
  transmissions: string[]
  drivetrains: string[]
  colors: string[]
  statuses: string[]
  prices: number[]
  price_min: number | null
  price_max: number | null
  mileage_min: number | null
  mileage_max: number | null
}

type ApiVehicleImagesResponse = {
  items: string[]
  total: number
  offset: number
  limit: number
  has_more: boolean
}

const apiOrigin = (() => {
  try {
    return new URL(apiClient.defaults.baseURL ?? '', window.location.origin).origin
  } catch {
    return window.location.origin
  }
})()

const fallbackVehicleImage = 'https://streetviewpixels-pa.googleapis.com/v1/thumbnail?cb_client=maps_sv.tactile&w=900&h=600&pitch=3.806508641649941&panoid=eMo88YOPWzb1MRZg7tLxVA&yaw=110.79005693573947'

const resolveImageUrl = (image: string) => {
  if (image.startsWith('/media/')) {
    return `${apiOrigin}${image}`
  }

  return image
}

const toVehicleDetails = (details?: ApiVehicle['details']): VehicleDetails => ({
  info: details?.info ?? [],
  sections: details?.sections ?? [],
  rawDescription: details?.raw_description,
})

const toVehicle = (vehicle: ApiVehicle): Vehicle => {
  const sourceImages = vehicle.images.length > 0 ? vehicle.images : [fallbackVehicleImage]
  const images = sourceImages.map(resolveImageUrl)

  return {
    id: vehicle.id,
    slug: vehicle.slug,
    year: vehicle.year,
    make: vehicle.make,
    model: vehicle.model,
    trim: vehicle.trim ?? '',
    price: vehicle.price,
    mileage: vehicle.mileage,
    bodyType: vehicle.body_type,
    transmission: vehicle.transmission ?? 'Automatic',
    drivetrain: vehicle.drivetrain ?? 'FWD',
    engine: vehicle.engine ?? 'Gasoline',
    exteriorColor: vehicle.exterior_color ?? 'Not listed',
    interiorColor: vehicle.interior_color ?? 'Not listed',
    vin: vehicle.vin ?? 'Not listed',
    stockNumber: vehicle.stock_number ?? vehicle.id,
    status: vehicle.status,
    shortDescription: vehicle.short_description,
    description: vehicle.description ?? vehicle.short_description,
    features: vehicle.features,
    specs: vehicle.specs ?? {},
    images,
    imagesTotal: vehicle.images_total || images.length,
    details: toVehicleDetails(vehicle.details),
    featured: vehicle.featured,
  }
}

const toListParams = (params: VehicleListParams) => ({
  page: params.page,
  page_size: params.pageSize,
  make: params.make || undefined,
  model: params.model || undefined,
  year_from: params.yearFrom,
  year_to: params.yearTo,
  body_type: params.bodyType || undefined,
  transmission: params.transmission || undefined,
  drivetrain: params.drivetrain || undefined,
  color: params.color || undefined,
  featured: params.featured,
  price_min: params.priceMin,
  price_max: params.priceMax,
  mileage_min: params.mileageMin,
  mileage_max: params.mileageMax,
  sort: params.sort,
})

export const listVehicles = async (params: VehicleListParams = {}) => {
  const response = await apiClient.get<ApiVehicleListResponse>('/vehicles', {
    params: toListParams(params),
  })

  return {
    items: response.data.items.map(toVehicle),
    total: response.data.total,
    page: response.data.page,
    pageSize: response.data.page_size,
  }
}

export const getVehicle = async (slug: string) => {
  const response = await apiClient.get<ApiVehicle>(`/vehicles/${slug}`, {
    params: { image_limit: 9 },
  })
  return toVehicle(response.data)
}

export const listVehicleImages = async (slug: string, offset = 0, limit = 8) => {
  const response = await apiClient.get<ApiVehicleImagesResponse>(`/vehicles/${slug}/images`, {
    params: { offset, limit },
  })

  return {
    items: response.data.items.map(resolveImageUrl),
    total: response.data.total,
    offset: response.data.offset,
    limit: response.data.limit,
    hasMore: response.data.has_more,
  }
}

export const getVehicleFilters = async (make?: string): Promise<VehicleFilters> => {
  const response = await apiClient.get<ApiVehicleFilters>('/vehicles/filters', {
    params: { make: make || undefined },
  })

  return {
    makes: response.data.makes,
    models: response.data.models,
    years: response.data.years,
    bodyTypes: response.data.body_types,
    transmissions: response.data.transmissions.filter(Boolean),
    drivetrains: response.data.drivetrains.filter(Boolean),
    colors: response.data.colors.filter(Boolean),
    statuses: response.data.statuses,
    prices: response.data.prices ?? [],
    priceMin: response.data.price_min,
    priceMax: response.data.price_max,
    mileageMin: response.data.mileage_min,
    mileageMax: response.data.mileage_max,
  }
}

File: backend/app/api/vehicles.py
from pathlib import Path

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.db.deps import get_db
from app.schemas.vehicle import (
    VehicleDetail,
    VehicleFilters,
    VehicleImagesResponse,
    VehicleListResponse,
)
from app.services.vehicle_service import (
    get_vehicle_by_slug,
    get_vehicle_filters,
    list_vehicles,
)

router = APIRouter(prefix="/vehicles", tags=["vehicles"])
MEDIA_ROOT = Path("app/static/media").resolve()


def media_file_exists(path: str) -> bool:
    if path.startswith("http://") or path.startswith("https://"):
        return True
    if not path.startswith("/media/"):
        return True

    relative_parts = path.removeprefix("/media/").split("/")
    file_path = (MEDIA_ROOT.joinpath(*relative_parts)).resolve()

    try:
        file_path.relative_to(MEDIA_ROOT)
    except ValueError:
        return False

    return file_path.is_file()


def existing_images(images: list[str] | None) -> list[str]:
    return [image for image in images or [] if media_file_exists(image)]


def public_vehicle_details(details: dict | None) -> dict:
    if not details:
        return {}

    public_details = dict(details)
    public_details["info"] = [
        item
        for item in details.get("info", [])
        if str(item.get("label", "")).strip().lower() != "vin"
    ]
    return public_details


def serialize_vehicle(vehicle, image_limit: int | None = None) -> dict:
    all_images = existing_images(vehicle.images)
    images = all_images[:image_limit] if image_limit is not None else all_images

    return {
        "id": vehicle.id,
        "slug": vehicle.slug,
        "title": vehicle.title,
        "make": vehicle.make,
        "model": vehicle.model,
        "trim": vehicle.trim,
        "year": vehicle.year,
        "status": vehicle.status,
        "stock_number": vehicle.stock_number,
        "vin": None,
        "price": vehicle.price,
        "mileage": vehicle.mileage,
        "body_type": vehicle.body_type,
        "transmission": vehicle.transmission,
        "drivetrain": vehicle.drivetrain,
        "engine": vehicle.engine,
        "exterior_color": vehicle.exterior_color,
        "interior_color": vehicle.interior_color,
        "location": vehicle.location,
        "short_description": vehicle.short_description,
        "images": images,
        "images_total": len(all_images),
        "features": vehicle.features or [],
        "specs": vehicle.specs or {},
        "details": public_vehicle_details(vehicle.details),
        "featured": vehicle.featured,
        "financing_available": vehicle.financing_available,
        "warranty_available": vehicle.warranty_available,
        "delivery_available": vehicle.delivery_available,
    }


@router.get("", response_model=VehicleListResponse)
def get_vehicle_list(
    page: int = Query(1, ge=1),
    page_size: int = Query(12, ge=1, le=50),
    make: str | None = None,
    model: str | None = None,
    year_from: int | None = Query(None, ge=1900),
    year_to: int | None = Query(None, ge=1900),
    body_type: str | None = None,
    transmission: str | None = None,
    drivetrain: str | None = None,
    color: str | None = None,
    status: str | None = None,
    featured: bool | None = None,
    q: str | None = Query(None, min_length=2),
    price_min: int | None = Query(None, ge=0),
    price_max: int | None = Query(None, ge=0),
    mileage_min: int | None = Query(None, ge=0),
    mileage_max: int | None = Query(None, ge=0),
    sort: str = Query("year_desc", pattern="^(year_desc|price_asc|price_desc|mileage_asc)$"),
    db: Session = Depends(get_db),
):
    items, total = list_vehicles(
        db=db,
        page=page,
        page_size=page_size,
        make=make,
        model=model,
        year_from=year_from,
        year_to=year_to,
        body_type=body_type,
        transmission=transmission,
        drivetrain=drivetrain,
        color=color,
        status=status,
        featured=featured,
        q=q,
        price_min=price_min,
        price_max=price_max,
        mileage_min=mileage_min,
        mileage_max=mileage_max,
        sort=sort,
    )

    return {
        "items": [serialize_vehicle(item, image_limit=1) for item in items],
        "total": total,
        "page": page,
        "page_size": page_size,
    }


@router.get("/filters", response_model=VehicleFilters)
def get_filters(make: str | None = None, db: Session = Depends(get_db)):
    return get_vehicle_filters(db, make=make)


@router.get("/{slug}/images", response_model=VehicleImagesResponse)
def get_vehicle_images(
    slug: str,
    offset: int = Query(0, ge=0),
    limit: int = Query(12, ge=1, le=48),
    db: Session = Depends(get_db),
):
    vehicle = get_vehicle_by_slug(db, slug)
    if not vehicle:
        raise HTTPException(status_code=404, detail="Vehicle not found")

    images = existing_images(vehicle.images)
    items = images[offset : offset + limit]
    next_offset = offset + len(items)
    return {
        "items": items,
        "total": len(images),
        "offset": offset,
        "limit": limit,
        "has_more": next_offset < len(images),
    }


@router.get("/{slug}", response_model=VehicleDetail)
def get_vehicle_detail(
    slug: str,
    image_limit: int = Query(12, ge=1, le=80),
    db: Session = Depends(get_db),
):
    vehicle = get_vehicle_by_slug(db, slug)
    if not vehicle:
        raise HTTPException(status_code=404, detail="Vehicle not found")
    payload = serialize_vehicle(vehicle, image_limit=image_limit)
    payload["description"] = vehicle.description
    payload["created_at"] = vehicle.created_at
    return payload
